semanticdb/semanticdb3/semanticdb3.md

```diff
@@ -351,35 +354,46 @@ references to nested type definitions, e.g. path-dependent types in Scala,
 type refs include `prefix`.
 
 ```protobuf
-message SingletonType {
-  enum Tag {
-    UNKNOWN_SINGLETON = 0;
-    SYMBOL = 1;
-    THIS = 2;
-    SUPER = 3;
-    UNIT = 4;
-    BOOLEAN = 5;
-    BYTE = 6;
-    SHORT = 7;
-    CHAR = 8;
-    INT = 9;
-    LONG = 10;
-    FLOAT = 11;
-    DOUBLE = 12;
-    STRING = 13;
-    NULL = 14;
-  }
-  Tag tag = 1;
-  Type prefix = 2;
-  string symbol = 3;
-  int64 primitive = 4;
-  string string = 5;
+message SingleType {
+  Type prefix = 1;
+  string symbol = 2;
+}
+```
+
+`SingleType` is a singleton type that is inhabited only by the value of
+the definition represented by `symbol` and the accompanying `prefix`.
+
+```protobuf
+message ThisType {
+  string symbol = 1;
 }
 ```
 
-`SingletonType` represents a singleton type. Singletons may be specified: 1)
-via a [Symbol](#symbol) accompanied with a `prefix`, 2) via
-a keyword (by `this` or `super`) or 3) via a literal.
+`ThisType` is a singleton type that is inhabited only by the value
+of `this` reference to the definition represented by `symbol`,
+only visible inside the lexical extent of the enclosing definition.
+
+```protobuf
+message SuperType {
+  Type prefix = 1;
+  string symbol = 2;
+}
+```
+
+`SuperType` is a singleton type that is inhabited only by the value
+of `super` reference to the definition represented by `symbol`,
+prefixed by an optional `prefix`, only visible inside the lexical extent
+of the enclosing definition.
+
+```protobuf
+message ConstantType {
+  Constant constant = 1;
```

Comment:
The spec is missing the definition of `Constant`

Reply:
Done. I meant to discuss whether the triviality of constants merits not including them in the spec, but it was simpler to just go ahead and add them.